File: Java/src/org/concentus/TonalityAnalysisState.java
/* Copyright (c) 2007-2008 CSIRO
   Copyright (c) 2007-2011 Xiph.Org Foundation
   Originally written by Jean-Marc Valin, Gregory Maxwell, Koen Vos,
   Timothy B. Terriberry, and the Opus open-source contributors
   Ported to Java by Logan Stromberg

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions
   are met:

   - Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.

   - Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.

   - Neither the name of Internet Society, IETF or IETF Trust, nor the
   names of specific contributors, may be used to endorse or promote
   products derived from this software without specific prior written
   permission.

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

package org.concentus;

class TonalityAnalysisState
{
    boolean enabled = false;
    final float[] angle = new float[240];
    final float[] d_angle = new float[240];
    final float[] d2_angle = new float[240];
    final int[] inmem = new int[OpusConstants.ANALYSIS_BUF_SIZE];
    int mem_fill;                      /* number of usable samples in the buffer */
    final float[] prev_band_tonality = new float[OpusConstants.NB_TBANDS];
    float prev_tonality;
    final float[][] E = Arrays.InitTwoDimensionalArray<float>(OpusConstants.NB_FRAMES, OpusConstants.NB_TBANDS);
    final float[] lowE = new float[OpusConstants.NB_TBANDS];
    final float[] highE = new float[OpusConstants.NB_TBANDS];
    final float[] meanE = new float[OpusConstants.NB_TOT_BANDS];
    final float[] mem = new float[32];
    final float[] cmean = new float[8];
    final float[] std = new float[9];
    float music_prob;
    float Etracker;
    float lowECount;
    int E_count;
    int last_music;
    int last_transition;
    int count;
    final float[] subframe_mem = new float[3];
    int analysis_offset;
    /** Probability of having speech for time i to DETECT_SIZE-1 (and music before).
        pspeech[0] is the probability that all frames in the window are speech. */
    final float[] pspeech = new float[OpusConstants.DETECT_SIZE];
    /** Probability of having music for time i to DETECT_SIZE-1 (and speech before).
        pmusic[0] is the probability that all frames in the window are music. */
    final float[] pmusic = new float[OpusConstants.DETECT_SIZE];
    float speech_confidence;
    float music_confidence;
    int speech_confidence_count;
    int music_confidence_count;
    int write_pos;
    int read_pos;
    int read_subframe;
    final AnalysisInfo[] info = new AnalysisInfo[OpusConstants.DETECT_SIZE];

    TonalityAnalysisState()
    {
        for (int c = 0; c < OpusConstants.DETECT_SIZE; c++)
        {
            info[c] = new AnalysisInfo();
        }
    }

    void Reset()
    {
        Arrays.MemSet<float>(angle,0, 240);
        Arrays.MemSet<float>(d_angle,0, 240);
        Arrays.MemSet<float>(d2_angle,0, 240);
        Arrays.MemSet(inmem, 0, OpusConstants.ANALYSIS_BUF_SIZE);
        mem_fill = 0;
        Arrays.MemSet<float>(prev_band_tonality,0, OpusConstants.NB_TBANDS);
        prev_tonality = 0;
        for (int c = 0; c < OpusConstants.NB_FRAMES; c++)
        {
            Arrays.MemSet<float>(E[c], 0, OpusConstants.NB_TBANDS);
        }
        Arrays.MemSet<float>(lowE,0, OpusConstants.NB_TBANDS);
        Arrays.MemSet<float>(highE,0, OpusConstants.NB_TBANDS);
        Arrays.MemSet<float>(meanE,0, OpusConstants.NB_TOT_BANDS);
        Arrays.MemSet<float>(mem,0, 32);
        Arrays.MemSet<float>(cmean,0, 8);
        Arrays.MemSet<float>(std,0, 9);
        music_prob = 0;
        Etracker = 0;
        lowECount = 0;
        E_count = 0;
        last_music = 0;
        last_transition = 0;
        count = 0;
        Arrays.MemSet<float>(subframe_mem,0, 3);
        analysis_offset = 0;
        Arrays.MemSet<float>(pspeech,0, OpusConstants.DETECT_SIZE);
        Arrays.MemSet<float>(pmusic,0, OpusConstants.DETECT_SIZE);
        speech_confidence = 0;
        music_confidence = 0;
        speech_confidence_count = 0;
        music_confidence_count = 0;
        write_pos = 0;
        read_pos = 0;
        read_subframe = 0;
        for (int c = 0; c < OpusConstants.DETECT_SIZE; c++)
        {
            info[c].Reset();
        }
    }
}

File: Go/opus/TonalityAnalysisState.go
package opus

type TonalityAnalysisState struct {
	enabled                 bool
	angle                   [240]float32
	d_angle                 [240]float32
	d2_angle                [240]float32
	inmem                   []int
	mem_fill                int
	prev_band_tonality      [NB_TBANDS]float32
	prev_tonality           float32
	E                       [NB_FRAMES][NB_TBANDS]float32
	lowE                    [NB_TBANDS]float32
	highE                   [NB_TBANDS]float32
	meanE                   [NB_TOT_BANDS]float32
	mem                     [32]float32
	cmean                   [8]float32
	std                     [9]float32
	music_prob              float32
	Etracker                float32
	lowECount               float32
	E_count                 int
	last_music              int
	last_transition         int
	count                   int
	subframe_mem            []float32
	analysis_offset         int
	pspeech                 [DETECT_SIZE]float32
	pmusic                  [DETECT_SIZE]float32
	speech_confidence       float32
	music_confidence        float32
	speech_confidence_count int
	music_confidence_count  int
	write_pos               int
	read_pos                int
	read_subframe           int
	info                    [DETECT_SIZE]*AnalysisInfo
}

func NewTonalityAnalysisState() TonalityAnalysisState {
	t := TonalityAnalysisState{}
	for i := 0; i < DETECT_SIZE; i++ {
		t.info[i] = &AnalysisInfo{}
	}
	return t
}

func (t *TonalityAnalysisState) Reset() {
	for i := range t.angle {
		t.angle[i] = 0
	}
	for i := range t.d_angle {
		t.d_angle[i] = 0
	}
	for i := range t.d2_angle {
		t.d2_angle[i] = 0
	}
	for i := range t.inmem {
		t.inmem[i] = 0
	}
	t.mem_fill = 0
	for i := range t.prev_band_tonality {
		t.prev_band_tonality[i] = 0
	}
	t.prev_tonality = 0
	for i := range t.E {
		for j := range t.E[i] {
			t.E[i][j] = 0
		}
	}
	for i := range t.lowE {
		t.lowE[i] = 0
	}
	for i := range t.highE {
		t.highE[i] = 0
	}
	for i := range t.meanE {
		t.meanE[i] = 0
	}
	for i := range t.mem {
		t.mem[i] = 0
	}
	for i := range t.cmean {
		t.cmean[i] = 0
	}
	for i := range t.std {
		t.std[i] = 0
	}
	t.music_prob = 0
	t.Etracker = 0
	t.lowECount = 0
	t.E_count = 0
	t.last_music = 0
	t.last_transition = 0
	t.count = 0
	for i := range t.subframe_mem {
		t.subframe_mem[i] = 0
	}
	t.analysis_offset = 0
	for i := range t.pspeech {
		t.pspeech[i] = 0
	}
	for i := range t.pmusic {
		t.pmusic[i] = 0
	}
	t.speech_confidence = 0
	t.music_confidence = 0
	t.speech_confidence_count = 0
	t.music_confidence_count = 0
	t.write_pos = 0
	t.read_pos = 0
	t.read_subframe = 0
	for i := 0; i < DETECT_SIZE; i++ {
		t.info[i].Reset()
	}
}
